File: utils_library/src/main/java/com/zeropercenthappy/utilslibrary/CacheUtils.java
package com.zeropercenthappy.utilslibrary;

import android.content.Context;
import android.os.Environment;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Locale;

/**
 * @author ybq
 */

public class CacheUtils {

    public static File getCacheDir(Context context) {
        File dir = context.getExternalCacheDir();
        if (dir == null) {
            dir = new File(Environment.getExternalStorageDirectory().getAbsoluteFile() + "/" + context.getPackageName());
        }
        if (!dir.exists()) {
            dir.mkdirs();
        }
        return dir;
    }

    public static boolean clearCacheDir(Context context) {
        File cacheDir = getCacheDir(context);
        return FileUtils.deleteFile(cacheDir);
    }

    public static File getImageCacheFile(Context context) {
        StringBuilder path = new StringBuilder();
        String cacheDirPath = getCacheDir(context).getAbsolutePath();
        path.append(cacheDirPath);
        path.append("/image/");
        File dir = new File(path.toString());
        if (!dir.exists()) {
            dir.mkdirs();
        }
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss", Locale.US);
        path.append(simpleDateFormat.format(System.currentTimeMillis()));
        path.append(".jpg");
        File file = new File(path.toString());
        try {
            if (!file.exists()) {
                file.createNewFile();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return file;
    }

    public static File getAudioCacheFile(Context context) {
        StringBuilder path = new StringBuilder();
        String cacheDirPath = getCacheDir(context).getAbsolutePath();
        path.append(cacheDirPath);
        path.append("/audio/");
        File dir = new File(path.toString());
        if (!dir.exists()) {
            dir.mkdirs();
        }
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss", Locale.US);
        path.append(simpleDateFormat.format(System.currentTimeMillis()));
        path.append(".mp3");
        File file = new File(path.toString());
        try {
            if (!file.exists()) {
                file.createNewFile();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return file;
    }

    public static File getVideoCacheFile(Context context) {
        StringBuilder path = new StringBuilder();
        String cacheDirPath = getCacheDir(context).getAbsolutePath();
        path.append(cacheDirPath);
        path.append("/video/");
        File dir = new File(path.toString());
        if (!dir.exists()) {
            dir.mkdirs();
        }
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss", Locale.US);
        path.append(simpleDateFormat.format(System.currentTimeMillis()));
        path.append(".mp4");
        File file = new File(path.toString());
        try {
            if (!file.exists()) {
                file.createNewFile();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return file;
    }

    public static File getLogCacheFile(Context context) {
        StringBuilder path = new StringBuilder();
        String cacheDirPath = getCacheDir(context).getAbsolutePath();
        path.append(cacheDirPath);
        path.append("/log/");
        File dir = new File(path.toString());
        if (!dir.exists()) {
            dir.mkdirs();
        }
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd_HH:mm:ss", Locale.US);
        path.append(simpleDateFormat.format(System.currentTimeMillis()));
        path.append(".log");
        File file = new File(path.toString());
        try {
            if (!file.exists()) {
                file.createNewFile();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return file;
    }
}

File: utilsLibrary/src/main/java/com/zeropercenthappy/utilslibrary/utils/CacheUtils.kt
package com.zeropercenthappy.utilslibrary.utils

import android.content.Context
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

/**
 * @author zeropercenthappy
 */

object CacheUtils {

    /**
     * 获取缓存文件夹
     * context.externalCacheDir ->  /storage/emulated/0/Android/data/{packageName}/cache
     * context.cacheDir -> /data/user/0/{packageName}/cache
     */
    @JvmStatic
    fun getCacheDir(context: Context): File {
        return context.externalCacheDir ?: context.cacheDir
    }

    /**
     * 获取缓存文件夹下的子文件夹
     */
    @JvmStatic
    fun getSubCacheDir(context: Context, subDirName: String): File {
        val dir = if (subDirName.isEmpty()) {
            getCacheDir(context)
        } else {
            File(getCacheDir(context), subDirName)
        }
        if (!dir.exists()) {
            dir.mkdirs()
        }
        return dir
    }

    @JvmStatic
    fun clearCacheDir(context: Context): Boolean {
        val dir = getCacheDir(context)
        return FileUtils.deleteFile(dir)
    }

    @JvmStatic
    fun clearSubCacheDir(context: Context, subDirName: String): Boolean {
        val dir = getSubCacheDir(context, subDirName)
        return FileUtils.deleteFile(dir)
    }

    /**
     * @param context
     *
     * @param fileName      想要保存的缓存文件名，包括后缀名
     *
     * @param createFile    若cacheFile不存在，是否需要自动创建空文件
     *
     * @return
     */
    @JvmStatic
    fun createCacheFile(context: Context,
                        fileName: String,
                        createFile: Boolean): File? {
        return createCacheFile(context, "", fileName, createFile)
    }

    /**
     * @param context
     *
     * @param subDirName    缓存存放的子目录
     *
     * @param fileName      想要保存的缓存文件名，包括后缀名
     *
     * @param createFile    若cacheFile不存在，是否需要自动创建空文件
     *
     * @return
     */
    @JvmStatic
    fun createCacheFile(context: Context,
                        subDirName: String,
                        fileName: String,
                        createFile: Boolean): File? {
        return try {
            val dir = getSubCacheDir(context, subDirName)
            val cacheFile = File(dir, fileName)
            if (createFile && !cacheFile.exists()) {
                cacheFile.createNewFile()
            }
            cacheFile
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    /**
     * @param context
     *
     * @param fileExtension 想要保存的缓存文件后缀名
     *
     * @param createFile    若cacheFile不存在，是否需要自动创建空文件
     *
     * @return
     */
    @JvmStatic
    fun createFormatCacheFile(
            context: Context,
            fileExtension: String,
            createFile: Boolean
    ): File? {
        return createFormatCacheFile(context, "", fileExtension, createFile)
    }

    /**
     * @param context
     *
     * @param subDirName       缓存存放的子目录
     *
     * @param fileExtension 想要保存的缓存文件后缀名
     *
     * @param createFile    若cacheFile不存在，是否需要自动创建空文件
     *
     * @return
     */
    @JvmStatic
    fun createFormatCacheFile(
            context: Context,
            subDirName: String,
            fileExtension: String,
            createFile: Boolean
    ): File? {
        return try {
            val simpleDateFormat = SimpleDateFormat("yyyyMMddHHmmss", Locale.US)
            val dir = getSubCacheDir(context, subDirName)
            val formattedExtension = if (fileExtension.startsWith(".")) fileExtension else ".$fileExtension"
            val cacheFile = File(dir, simpleDateFormat.format(System.currentTimeMillis()) + formattedExtension)
            if (createFile && !cacheFile.exists()) {
                cacheFile.createNewFile()
            }
            cacheFile
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}
